### DISCUSSION TITLE
database design suggestion

### DISCUSSION THREAD

**Discussion by cduytran** [pocketbase/pocketbase#1801] [Q&A]
My current design is this:

Collection pack has two fields
- packNumber: text
- packTickets : JSON [ {   }, {  } ... ] - the object inside this array holds { price: number, soldDate: text }

I am thinking about changing the design to:

Collection pack will still have only one field
-packNumber: text

Then, I have another collection _ticket_ which has the following fields:
-price: number,
-soldDate: date
-packRel: a relation field that references pack

Is there any performance differences between the two implementations? Is one way better than another? Thanks.

**Comment by ganigeorgiev** [comment]
It depends on your desired workflow and data access patterns.

I personally will choose the 2 separate collections structure, because:

- You have field value type normalizations and constraints (`price` will be always number, `soldDate` will be always a datetime in the format `Y-m-d H:i:s.uZ`, etc.).
The `json` field doesn't have any schema validators and it can be anything.

- Keeping the tickets in their own collection will allow you to perform search queries server-side across all `tickets` records, eg. `price > 10`.
The same cannot be done done with the `json` field, because you'll have to specify the individual array index before accessing the object key, eg. `packTickets.0.price > 10`.

- You have more granular control over who can access and modify the tickets since they will have their own API rules.

**Reply by cduytran** [reply]
So, I am building an inventory system that keeps tracks of receiving and selling tickets that have an unique package number. 

Each day, we receive thousand of packages with an unique pack number. And each package has hundreds of tickets inside marked with a string of text in this format:  packnumber-ticketnumber ( for an example 1234-0001 where 1234 is the pack number and 0001 is the ticket number).  That's why I thought of holding each ticket inside an array and use the index as the ticket number.

pack.packTickets[0] is the first ticket inside a pack.

I'd like to make it so that at any point I can query the database for number of tickets sold on certain date, how many tickets we have on-hand, etc.. standard database stuff I guess. Any suggestion?

**Reply by ganigeorgiev** [reply]
> I'd like to make it so that at any point I can query the database for number of tickets sold on certain date, how many tickets we have on-hand, etc.. standard database stuff I guess. Any suggestion?

I think  the separate `tickets` collection approach will be better for your use case, because as mentioned in the previous post, it will allow you to perform complex search server-side queries across all tickets and this cannot be achieved with the `json` Array<Object> field (at least not directly).

Also note that non-count aggregations are currently not supported out of the box (aka. the web APIs doesn't provide a mechanism to perform sum/group by/etc. queries) but this will be available in the next v0.13.0 release.
